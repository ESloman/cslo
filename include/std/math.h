/**
 * @file math.h
 * @brief Header file for math-related functions and constants.
 */

#ifndef cslo_std_math_h
#define cslo_std_math_h

#include "core/object.h"
#include "core/value.h"

/**
 * @brief Gets the math module with all its functions.
 * @return A pointer to the ObjDict containing math functions.
 */
ObjModule* getMathModule();

/**
 * Calculates the ceil of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value ceilNative(int argCount, Value* args);

/**
 * Calculates the floor of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value floorNative(int argCount, Value* args);

/**
 * Calculates the square root of a number.
 */
Value sqrtNative(int argCount, Value* args);

/**
 * Calculates the sine of a number.
 */
Value sinNative(int argCount, Value* args);

/**
 * Calculates the cosine of a number.
 */
Value cosNative(int argCount, Value* args);

/**
 * Calculates the tangent of a number.
 */
Value tanNative(int argCount, Value* args);


#endif  // cslo_std_math_h
