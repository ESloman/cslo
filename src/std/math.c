/**
 * @file math.c
 * @brief Implementation of math-related functions and constants.
 */

#include <math.h>

#include "core/value.h"
#include "std/math.h"

/**
 * @brief Gets the math module with all its functions.
 * @return A pointer to the ObjDict containing math functions.
 */
ObjModule* getMathModule() {
    ObjModule* mathModule = newModule();
    tableSet(&mathModule->methods, OBJ_VAL(copyString("ceil", 4)), OBJ_VAL(newNative(ceilNative)));
    tableSet(&mathModule->methods, OBJ_VAL(copyString("floor", 5)), OBJ_VAL(newNative(floorNative)));
    tableSet(&mathModule->methods, OBJ_VAL(copyString("sqrt", 4)), OBJ_VAL(newNative(sqrtNative)));
    tableSet(&mathModule->methods, OBJ_VAL(copyString("sin", 3)), OBJ_VAL(newNative(sinNative)));
    tableSet(&mathModule->methods, OBJ_VAL(copyString("cos", 3)), OBJ_VAL(newNative(cosNative)));
    tableSet(&mathModule->methods, OBJ_VAL(copyString("tan", 3)), OBJ_VAL(newNative(tanNative)));
    return mathModule;
}

/**
 * Calculates the ceil of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value ceilNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    return NUMBER_VAL(ceil(value));
}

/**
 * Calculates the floor of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value floorNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    return NUMBER_VAL(floor(value));
}

/**
 * Calculates the square root of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value sqrtNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    if (value < 0) {
        return ERROR_VAL;
    }
    return NUMBER_VAL(sqrt(value));
}

/**
 * Calculates the sine of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value sinNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    return NUMBER_VAL(sin(value));
}

/**
 * Calculates the cosine of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value cosNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    return NUMBER_VAL(cos(value));
}

/**
 * Calculates the tangent of a number.
 * If the argument is not a number, returns NIL_VAL or throws an error.
 */
Value tanNative(int argCount, Value* args) {
    if (argCount != 1 || !IS_NUMBER(args[0])) {
        return ERROR_VAL;
    }
    double value = AS_NUMBER(args[0]);
    return NUMBER_VAL(tan(value));
}
