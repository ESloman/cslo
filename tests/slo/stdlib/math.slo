
import math;

# ceil and floor
print("Printing ceil and floor stuff");
for (var x = 0; x < 10; x = x + 1.1) {
    print("Testing with ", x, "; ceil: ", math.ceil(x), ", floor: ", math.floor(x));
}

# sqrt
print(math.sqrt(9));  # 3

# sin
print(math.sin(1));

# cos
print(math.cos(1));

# tan
print(math.tan(1));
