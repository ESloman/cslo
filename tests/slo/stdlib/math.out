Printing ceil and floor stuff
Testing with 0; ceil: 0, floor: 0
Testing with 1.1; ceil: 2, floor: 1
Testing with 2.2; ceil: 3, floor: 2
Testing with 3.3; ceil: 4, floor: 3
Testing with 4.4; ceil: 5, floor: 4
Testing with 5.5; ceil: 6, floor: 5
Testing with 6.6; ceil: 7, floor: 6
Testing with 7.7; ceil: 8, floor: 7
Testing with 8.8; ceil: 9, floor: 8
Testing with 9.9; ceil: 10, floor: 9
3
0.841471
0.540302
1.55741
